In [5]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

#For data visualisations
import matplotlib.pyplot as plt
import seaborn as sns

In [6]:
#reading df file
df=pd.read_csv("file_02.csv")

In [7]:
#printing first 10 rows of the dataset
df.head(10)

index        Date        Region Thermal Generation Actual (in MU)  \
0      0  2017-09-01      Northern                            624.23   
1      1  2017-09-01       Western                          1,106.89   
2      2  2017-09-01      Southern                            576.66   
3      3  2017-09-01       Eastern                            441.02   
4      4  2017-09-01  NorthEastern                             29.11   
5      5  2017-09-02      Northern                            624.23   
6      6  2017-09-02       Western                          1,106.89   
7      7  2017-09-02      Southern                            576.66   
8      8  2017-09-02       Eastern                            441.02   
9      9  2017-09-02  NorthEastern                             29.11   

  Thermal Generation Estimated (in MU)  Nuclear Generation Actual (in MU)  \
0                               484.21                              30.36   
1                             1,024.33                              25.17   
2                               578.55                              62.73   
3                               429.39                                NaN   
4                                15.91                                NaN   
5                               507.42                              30.36   
6                             1,050.91                              25.17   
7                               562.79                              62.73   
8                               425.75                                NaN   
9                                16.50                                NaN   

   Nuclear Generation Estimated (in MU)  Hydro Generation Actual (in MU)  \
0                                 35.57                           273.27   
1                                  3.81                            72.00   
2                                 49.80                           111.57   
3                                   NaN                            85.94   
4                                   NaN                            24.64   
5                                 35.69                           273.27   
6                                  3.83                            72.00   
7                                 52.76                           111.57   
8                                   NaN                            85.94   
9                                   NaN                            24.64   

   Hydro Generation Estimated (in MU)  
0                              320.81  
1                               21.53  
2                               64.78  
3                               69.36  
4                               21.21  
5                              317.19  
6                               23.97  
7                               59.49  
8                               74.18  
9                               20.81

In [8]:
#printing last 10 rows of the dataset
df.tail(10)

index        Date        Region Thermal Generation Actual (in MU)  \
4935    300  2020-07-31      Northern                            708.11   
4936    301  2020-07-31       Western                          1,199.28   
4937    302  2020-07-31      Southern                            543.60   
4938    303  2020-07-31       Eastern                            496.03   
4939    304  2020-07-31  NorthEastern                             36.98   
4940    305  2020-08-01      Northern                            669.47   
4941    306  2020-08-01       Western                          1,116.00   
4942    307  2020-08-01      Southern                            494.66   
4943    308  2020-08-01       Eastern                            482.86   
4944    309  2020-08-01  NorthEastern                             34.42   

     Thermal Generation Estimated (in MU)  Nuclear Generation Actual (in MU)  \
4935                               625.88                              26.88   
4936                             1,288.65                              42.37   
4937                               444.43                              61.83   
4938                               553.21                                NaN   
4939                                30.94                                NaN   
4940                               602.96                              26.88   
4941                             1,262.10                              42.37   
4942                               415.53                              61.83   
4943                               547.03                                NaN   
4944                                32.81                                NaN   

      Nuclear Generation Estimated (in MU)  Hydro Generation Actual (in MU)  \
4935                                 22.78                           327.08   
4936                                 36.68                            29.75   
4937                                 26.21                            61.43   
4938                                   NaN                            80.00   
4939                                   NaN                            27.51   
4940                                 23.41                           348.72   
4941                                 36.63                            54.67   
4942                                 26.28                            93.49   
4943                                   NaN                            87.22   
4944                                   NaN                            29.87   

      Hydro Generation Estimated (in MU)  
4935                              348.39  
4936                               19.26  
4937                               85.42  
4938                               96.30  
4939                               29.76  
4940                              351.98  
4941                               20.28  
4942                               77.25  
4943                               93.78  
4944                               29.96

From the code it can seen that there are some null values in the data.In Eastern zone and North Eastern Zone mainly the null values are there for Nuclear Power generation.IT is because there are no nuclear power plant in Eastern and North Eastern India.

Data is taken from 1st of september 2017 -1st of august 2020

In [9]:
#getting info of df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4945 entries, 0 to 4944
Data columns (total 9 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   index                                 4945 non-null   int64  
 1   Date                                  4945 non-null   object 
 2   Region                                4945 non-null   object 
 3   Thermal Generation Actual (in MU)     4945 non-null   object 
 4   Thermal Generation Estimated (in MU)  4945 non-null   object 
 5   Nuclear Generation Actual (in MU)     2967 non-null   float64
 6   Nuclear Generation Estimated (in MU)  2967 non-null   float64
 7   Hydro Generation Actual (in MU)       4945 non-null   float64
 8   Hydro Generation Estimated (in MU)    4945 non-null   float64
dtypes: float64(4), int64(1), object(4)
memory usage: 347.8+ KB


There are 4945 rows and 9 columns are there.
Only  Nuclear Generation Actual (in MU) and  Nuclear Generation Estimated (in MU) have null values.
There are 4 float datatypes,1 integer and 4 object datatype variables.

In [10]:
#Replacing delimiter in Values in Thermal generation Actual and Estimated
df["Thermal Generation Actual (in MU)"]=df["Thermal Generation Actual (in MU)"].str.replace(',','')
df["Thermal Generation Estimated (in MU)"]=df["Thermal Generation Estimated (in MU)"].str.replace(',','')

In [11]:
#converting Thermal Generation Actual(in MU) and Thermal Generation Actual(in MU) from object to float
df["Thermal Generation Actual (in MU)"] = df["Thermal Generation Actual (in MU)"].astype(float)
df["Thermal Generation Estimated (in MU)"] = df["Thermal Generation Estimated (in MU)"].astype(float)

In [12]:
#getting statistical description of the data
df.describe()

index  Thermal Generation Actual (in MU)  \
count  4945.000000                        4945.000000   
mean   2181.433771                         603.978358   
std    1397.754755                         383.534208   
min       0.000000                          12.340000   
25%     926.000000                         470.050000   
50%    2162.000000                         615.280000   
75%    3398.000000                         689.530000   
max    4634.000000                        1395.970000   

       Thermal Generation Estimated (in MU)  \
count                           4945.000000   
mean                             575.395116   
std                              383.387299   
min                               12.380000   
25%                              427.460000   
50%                              535.980000   
75%                              672.740000   
max                             1442.380000   

       Nuclear Generation Actual (in MU)  \
count                        2967.000000   
mean                           37.242208   
std                            15.883968   
min                             0.000000   
25%                            26.140000   
50%                            30.720000   
75%                            46.830000   
max                            68.740000   

       Nuclear Generation Estimated (in MU)  Hydro Generation Actual (in MU)  \
count                           2967.000000                      4945.000000   
mean                              36.987877                        73.305921   
std                               11.491292                        74.482145   
min                                0.000000                         0.000000   
25%                               30.190000                        26.910000   
50%                               34.840000                        52.960000   
75%                               43.075000                        85.940000   
max                               76.640000                       348.720000   

       Hydro Generation Estimated (in MU)  
count                         4945.000000  
mean                            76.842965  
std                             82.043952  
min                              0.000000  
25%                             23.310000  
50%                             50.270000  
75%                             95.800000  
max                            397.380000

Observations:1)Thermal power actual generation mean is more than estimated generation mean.Same with nuclear power generation.
2)Where as Hydro Power generation in actual is less than estimated hydro power generation.
3)Min power generation :-Thermal-12.34 MU,Nuclear-0,Hydro -0
4)Max power generation :-Thermal-1395.97 MU,Nuclear-68.74 MU,Hydro-348.72 MU

In [13]:
#getting unique values from region columns
df['Region'].unique()

array(['Northern', 'Western', 'Southern', 'Eastern', 'NorthEastern'],
      dtype=object)

There are 5 region in the dataset-'Northern', 'Western', 'Southern', 'Eastern', 'NorthEastern'

In [14]:
#Counting duplicate values
df.duplicated().sum()

0

There are no duplicate values

In [15]:
#Counting null value
df.isnull().sum()

index                                      0
Date                                       0
Region                                     0
Thermal Generation Actual (in MU)          0
Thermal Generation Estimated (in MU)       0
Nuclear Generation Actual (in MU)       1978
Nuclear Generation Estimated (in MU)    1978
Hydro Generation Actual (in MU)            0
Hydro Generation Estimated (in MU)         0
dtype: int64

There are 1978 null values both in Nuclear Generation Actual (MU) and Nuclear Generation Estimated(MU)

In [16]:
#Counting null values according to Region
missing_pct=df.isnull().sum()*100/len(df)
missing_pct

index                                    0.0
Date                                     0.0
Region                                   0.0
Thermal Generation Actual (in MU)        0.0
Thermal Generation Estimated (in MU)     0.0
Nuclear Generation Actual (in MU)       40.0
Nuclear Generation Estimated (in MU)    40.0
Hydro Generation Actual (in MU)          0.0
Hydro Generation Estimated (in MU)       0.0
dtype: float64

40 % of null values are present in Nuclear generation column

In [17]:
df.groupby(['Region']).count()

index  Date  Thermal Generation Actual (in MU)  \
Region                                                         
Eastern         989   989                                989   
NorthEastern    989   989                                989   
Northern        989   989                                989   
Southern        989   989                                989   
Western         989   989                                989   

              Thermal Generation Estimated (in MU)  \
Region                                               
Eastern                                        989   
NorthEastern                                   989   
Northern                                       989   
Southern                                       989   
Western                                        989   

              Nuclear Generation Actual (in MU)  \
Region                                            
Eastern                                       0   
NorthEastern                                  0   
Northern                                    989   
Southern                                    989   
Western                                     989   

              Nuclear Generation Estimated (in MU)  \
Region                                               
Eastern                                          0   
NorthEastern                                     0   
Northern                                       989   
Southern                                       989   
Western                                        989   

              Hydro Generation Actual (in MU)  \
Region                                          
Eastern                                   989   
NorthEastern                              989   
Northern                                  989   
Southern                                  989   
Western                                   989   

              Hydro Generation Estimated (in MU)  
Region                                            
Eastern                                      989  
NorthEastern                                 989  
Northern                                     989  
Southern                                     989  
Western                                      989

Observations:only 2 region has null values ,Eastern and North Eastern region,where there are no nuclear power plants

In [18]:
df1=df.groupby(by='Region')[['Thermal Generation Actual (in MU)','Thermal Generation Estimated (in MU)','Nuclear Generation Actual (in MU)','Nuclear Generation Estimated (in MU)','Hydro Generation Actual (in MU)','Hydro Generation Estimated (in MU)']].sum()
df1

Thermal Generation Actual (in MU)  \
Region                                            
Eastern                               482123.72   
NorthEastern                           32115.79   
Northern                              655048.26   
Southern                              610753.56   
Western                              1206631.65   

              Thermal Generation Estimated (in MU)  \
Region                                               
Eastern                                  468831.45   
NorthEastern                              33105.95   
Northern                                 602168.52   
Southern                                 543828.71   
Western                                 1197394.22   

              Nuclear Generation Actual (in MU)  \
Region                                            
Eastern                                    0.00   
NorthEastern                               0.00   
Northern                               26964.08   
Southern                               55855.21   
Western                                27678.34   

              Nuclear Generation Estimated (in MU)  \
Region                                               
Eastern                                       0.00   
NorthEastern                                  0.00   
Northern                                  31378.94   
Southern                                  46483.24   
Western                                   31880.85   

              Hydro Generation Actual (in MU)  \
Region                                          
Eastern                              48686.62   
NorthEastern                         17612.32   
Northern                            188854.16   
Southern                             71109.34   
Western                              36235.34   

              Hydro Generation Estimated (in MU)  
Region                                            
Eastern                                 52461.95  
NorthEastern                            14058.66  
Northern                               201204.03  
Southern                                77083.80  
Western                                 35180.02